rclpy/rclpy/action/server.py

```diff
@@ -602,16 +596,8 @@ def register_execute_callback(self, execute_callback):
 
     def destroy(self):
         """Destroy the underlying action server handle."""
-        if self._handle is None:
-            return
-
         for goal_handle in self._goal_handles.values():
             goal_handle.destroy()
 
         self._handle.destroy_when_not_in_use()
```

Comment:
```suggestion
        with self._node.handle:
            self._handle.destroy_when_not_in_use()
```

is needed to call the following?

https://github.com/ros2/rclpy/blob/53771be14430b360a7efb62bdd64085b25bb8707/rclpy/src/rclpy/action_server.cpp#L32

Reply:
Hm, good question.  When I was discussing this with @sloretz , he suggested that it wasn't necessary here, but I'm honestly not sure.  @sloretz can you comment?

Reply:
The action server calling `node_.destroy()` is surprising to me. I'll have a closer look.

Reply:
There's no need to call `with self._node.handle` here, even if ``destroy_when_not_in_use()` ends up calling `ActionServer.destroy()` right away.

It's a little weird because in rclpy there's a desire both to have the option to destroy rcl objects ASAP and the requirement to gracefully garbage collect them when no one explicitly destroys them. What's happening is `ActionServer` has a copy of a `Node` object and a shared pointer to the `rcl_action_server_t` type. All the classes have this pattern: Have a copy of the type they need kept alive (`ActionServer` needs `Node` to be alive for it to work), and a shared pointer to the rcl type. The shared pointer has a destructor that free's the `rcl` type.

Destroying objects right away is done by releasing shared pointers to the rcl types from the bottom up: `ActionServer` releases the reference to the `rcl_action_server_t` type before asking `Node` to release the reference to it's type. Garbage collection works by keeping those shared pointers alive from the bottom up (`ActionServer` has a `Node` so that `ActionServer` will be garbage collected before `Node`).  The context manager for each says "I'm using the rcl functions of this thing right now, so don't destroy it please". `node_.destroy()` isn't using the `rcl_node_t` instance, it's getting rid of the reference to it so it never uses it again. That's why `with self._node.handle:` doesn't need to be called.

Reply:
@sloretz thanks for the note!